integration/simulation/simulation_test.go

```diff
@@ -7,84 +7,176 @@ import (
+		avgBlockDurationUSecs,
+		avgGossipPeriod,
+		stats,
+	)
+
+	// execute the simulation
+	simulation.Start(txManager, simulationTimeSecs)
+
+	// run tests
+	checkBlockchainValidity(t, txManager, simulation)
+
+	t.Logf("%+v\n", stats)
+	// pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+}
+
+func checkBlockchainValidity(t *testing.T, txManager *TransactionManager, network *Simulation) {
+	// TODO check all nodes are the same height ?
+	// pick one node to draw height
+	l1Node := network.l1Network.nodes[0]
+	obscuroNode := network.l2Network.nodes[0]
+	currentBlockHead := obscuroNode.Headers().GetCurrentBlockHead()
+	currentRollupHead := obscuroNode.Headers().GetCurrentRollupHead()
+
+	l1Height := currentBlockHead.Height
+	l1HeightHash := currentBlockHead.ID
+	l2Height := currentRollupHead.Height
+
+	// ensure the L1 blocks are valid
+	validateL1(t, network.l1Network.Stats, l1Height, &l1HeightHash, l1Node)
+
+	// ensure the validity of l1 vs l2 stats
+	validateL1L2Stats(t, obscuroNode, network.l1Network.Stats)
+
+	// ensure the generated withdrawal stats match the l2 blockchain state (withdrawals)
+	totalWithdrawn := validateL2WithdrawalStats(t, obscuroNode, network.l1Network.Stats, l2Height, txManager)
+
+	// ensure that each node has the expected total balance computed above
+	validateL2NodeBalances(t, network.l2Network, network.l1Network.Stats, totalWithdrawn, txManager.wallets)
+
+	// ensure that each node can fetch each of the generated transactions
+	validateL2TxsExist(t, network.l2Network.nodes, txManager)
+}
+
+// validateL1L2Stats validates blockchain wide properties between L1 and the L2
+func validateL1L2Stats(t *testing.T, node *obscuro_node.Node, stats *Stats) {
+	l1HeaderCount := uint(0)
+	for header := node.Headers().GetCurrentBlockHead(); header != nil; header = node.Headers().GetBlockHeader(header.Parent) {
+		l1HeaderCount++
 	}
-	f, err := os.CreateTemp("../.build/simulations", "simulation-result-*.txt")
-	if err != nil {
-		panic(err)
+	l2HeaderCount := uint(0)
+	for header := node.Headers().GetCurrentRollupHead(); header != nil; header = node.Headers().GetRollupHeader(header.Parent) {
+		l2HeaderCount++
 	}
-	defer f.Close()
-	common.SetLog(f)
 
-	blockDuration := uint64(20_000)
-	l1netw, l2netw := RunSimulation(5, 10, 15, blockDuration, blockDuration/15, blockDuration/3)
-	firstNode := l2netw.nodes[0]
-	checkBlockchainValidity(t, l1netw, l2netw, firstNode.Enclave.TestDB(), firstNode.Enclave.TestPeekHead().Head)
-	stats := l1netw.Stats
-	fmt.Printf("%+v\n", stats)
-	// pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+	if l1HeaderCount > stats.totalL1Blocks || l2HeaderCount > stats.totalL2Blocks {
+		t.Errorf("should not have more blocks/rollups in stats than in the node header "+
+			"- Blocks: Header %d, Stats %d - Rollups: Header %d, Stats %d ",
+			l1HeaderCount,
+			stats.totalL1Blocks,
+			l2HeaderCount,
+			stats.totalL2Blocks,
+		)
+	}
+
+	efficiency := float64(l1HeaderCount-l2HeaderCount) / float64(l1HeaderCount)
```

Comment:
You should create a new constant for this. Like: `L2ToL1EfficiencyThreashold`

Reply:
I wasn't sure what efficiency should be, so just reused one 😄